### ISSUE TITLE
Many thrown errors have the same error data

### ISSUE THREAD

**Issue by ryanvanderpol** [expo/expo#307]
Almost every time I catch an error and write it to the console, it just prints out the following:

```
{"line":5311,"column":14,"sourceURL":"http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles"}
```

This is clearly not the actual error.  I'm not quite sure what I'm doing wrong.  This seems to happen more frequently in callbacks or in Promises.  It's very frustrating because it makes it nearly impossible to track down where the actual error is.  I've occasionally been able to get a real error message by wrapping some stuff in `setTimeout` calls, but I haven't found that to be consistent yet.

I'm pretty confident this is a problem with Expo, although given the message content it could be something internal to CRNA.  Any ideas here?

**Comment by dikaiosune** [comment]
We do override the `console.*` methods in order to send the messages back to your terminal. Do you think you could provide a reproduction of this issue using https://snack.expo.io? It'd really help to see some concrete code that triggers this behavior. Thanks!

**Comment by ryanvanderpol** [comment]
@dikaiosune I certainly can try.  How do I see `console.log` statements in Snack?

**Comment by ryanvanderpol** [comment]
I found it.  I've tried replicating the issue in this Snack: https://snack.expo.io/SyRDoZ4Eb

What I'm seeing in there is not the exact same problem, but it's still a problem.  In the `render` function I create a new Promise and inside that promise some code throws an error.  In this case I would expect the error written to the logs to say `ReferenceError: Animated is not defined`.  Instead, I get the following.  Removing the Promise gives me the behavior I expect.

This is actually more helpful than what I'm seeing in my real project because this message actually says `render` in there so I could probably figure out that it's coming from the `render` function, but it still doesn't tell me the error message or give me an accurate stack trace.

```
f@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:183
d@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:1575
c@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:431
render
https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:1696:3146
_renderValidatedComponentWithoutOwnerOrContext@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:8615
_renderValidatedComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:8782
performInitialMount@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:3114
mountComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:2167
mountComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:174:181
_updateRenderedComponentWithNextElement@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:8262
_updateRenderedComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:7831
_performComponentUpdate@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:7223
updateComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:6288
receiveComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:4973
receiveComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:174:581
_updateRenderedComponentWithNextElement@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:7983
_updateRenderedComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:7831
_performComponentUpdate@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:7223
updateComponent@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:6288
performUpdateIfNecessary@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:181:5200
performUpdateIfNecessary@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:174:945
s@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:178:832
perform@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:179:596
perform@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:179:596
U@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:178:1703
U@[native code]
closeAll@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:179:1249
perform@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:179:683
l@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:178:1023
u@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:177:65
enqueueSetState@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:177:987
setState@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:75:3078
https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:13:4417
n@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:232
https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:1649
n@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:232
r@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:489
https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:780
f@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:183
d@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:1575
c@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:37:431
n@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:764
e@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:816
async@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:124:3306
_updateCodeNotDebounced@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:13:4358
r@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:867:98
s@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:867:383
l@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:867:316
callTimer@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:22:626
callTimers@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:22:1010
value@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:20:2893
https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:20:832
value@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:20:2294
value@https://d1wp6m56sqw74a.cloudfront.net:443/%40exponent%2Fsnack%2F1.0.0%2F1ae38fba96975ef6d05bb5aefccc0920-18.0.0-ios.js:20:804
value@[native code]
```

**Comment by ide** [comment]
I suspect that somewhere inside the React Native internals there is code that catches your error and then rethrows a new one (which is why you're seeing the same line & col information). A small repro case (even not on Snack) would be helpful. In the meantime, if you log `error.message` or `error.stack` does that help unblock you?

**Comment by ryanvanderpol** [comment]
@ide I've been battling finding the source of one of these errors for the last hour or so, so your tip was timely.  Logging `error.message` actually does help a bit (I discovered there was a JSON parsing error via the message this time).

`error.stack` has the following, which isn't very helpful.

```
[exp] tryCallOne@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:5311:14
[exp] http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:5397:25
[exp] callTimer@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3633:17
[exp] callImmediatesPass@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3719:36
[exp] callImmediates@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3729:48
[exp] __guard@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3488:11
[exp] __callImmediates@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3499:19
[exp] http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3375:31
[exp] __guard@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3488:11
[exp] callFunctionReturnFlushedQueue@http://packager.4y-r7u.xxx.yyy.exp.direct:80/node_modules/react-native-scripts/build/bin/crna-entry.bundle?platform=ios&dev=true&strict=false&minify=false&hot=false&assetPlugin=expo/tools/hashAssetFiles:3373:19
[exp] callFunctionReturnFlushedQueue@[native code]
```

**Comment by ide** [comment]
Another thing you could do (not sure if it works for your use case though) is not catch the error and install a global error handler instead (example: https://github.com/expo/sentry-expo/blob/master/index.js#L35).

The downside of this is that your app will crash because of the uncaught error, but if your goal is to identify crashes anyway, this might work for you.

**Comment by ryanvanderpol** [comment]
Cool, yeah, not ideal but I'll set that up, too.

**Comment by terribleben** [comment]
Brent fixed this in [SDK 21](https://blog.expo.io/expo-sdk-21-0-0-is-now-available-be33b79921b7), closing.
